Refactor/irstate attrs

gorsing opened the pull request:
Performance: pragma(inline, false) hints the optimizer to inline a hot, side-effect-free predicate used in nearly every expression lowering.

Benefits

    Zero functional change; external API and output are untouched.

    Slightly smaller/faster IR generation in tight loops.

    Lays groundwork: other trivial helpers can now be upgraded in follow-ups.

gorsing committed:
add ddoc for some func

gorsing committed:
rewrite

gorsing committed:
fix after review

gorsing committed:
fix after review

gorsing committed:
fix after review

gorsing committed:
Merge branch 'dlang:master' into master

gorsing committed:
Merge branch 'dlang:master' into master

gorsing committed:
Merge branch 'dlang:master' into master

gorsing committed:
refactor/irstate-attrs

refactoring

add pragma(inline, true)

change pragma(inline, true) -> pragma(inline, false)

introduce root.Array.only and test it out in expressionsem (#21520)

[attrib.d] remove unused semantic import

[dstruct.d] remove unused import

[aggregate.d] move `searchCtor` to `expressionsem.d`

remove some now unused semantic symbols from the import list.

[cond.d] remove unused `typesem` import (#21538)

[expression.d] restrict `typesem` import

fix #19587 No debug line info for simple code blocks (#21544)

set location info on return statement, not only on return expression

[dsymbol.d] remove dependance on glue layer (#21532)

move `loadModuleFromLibrary` to `dsymbolsem.d` (#21535)

remove `dmodule.d` dependance of a few more `dsymbolsem.d` symbols.

[func.d] remove dependance on glue layer (#21534)

[aggregate.d] move `checkOverlappedFields` to `dsymbolsem.d` (#21542)

and make private

Templatize `_adEq2` (#21513)

* Mark array literals as `on-stack` to bypass `@nogc` analysis

* Update `fail_compilation` test output

After updating the memcmp-ability criteria, now dynamic array can also
be compared using `memcmp` if their element type is comparable bit by
bit.

refactor: use new array constructor throughout expressionsem

refactoring

gorsing committed:
refactor/irstate-attrs

gorsing committed:
refactoring

gorsing committed:
add pragma(inline, true)

gorsing committed:
change pragma(inline, true) -> pragma(inline, false)

gorsing committed:
Merge branch 'dlang:master' into refactor/irstate-attrs

ibuclaw reviewed (changes requested):
> Performance: pragma(inline, false) hints the optimizer to inline a hot, side-effect-free predicate used in nearly every expression lowering.

This statement is false, so I question whether this refactor is really needed.

gorsing commented:
Using pragma(inline, true) triggers errors in the Bootstrap tests.

I try pragma(inline, false)
I may have misunderstood its behavior—I assumed the compiler would still optimize where it can. We can safely drop the pragma(inline, false).

gorsing committed:
refactoring

gorsing commented:
Could I suggest you consider keeping

```d
@property
@safe pure nothrow @nogc
```

as it makes the code semantically cleaner?

gorsing committed:
Merge branch 'refactor/irstate-attrs' of github.com:gorsing/dmd into refactor/irstate-attrs

ibuclaw commented:
`pragma(inline, false)` _disables_ inlining. Calls can still be optimized away via other means, just not if the call is considered live.

Adding attributes to functions won't be moving any dials wrt code improvements.

Generally accepted refactorings tend to fall under the following categories.
- Reducing deeply nested branches with early returns.
- Replacing complex uses of `goto` with functions - toplevel or nested depending on whether the jump is to retry with different parameters or  jump to the exit branch.
- Making functions `private` by moving them to the one module where they are used.
- Reduce number of imports, simplifying the dependency graph between modules.
- Reduce the total lines of code by way of replacing duplicate code or a clunky idiom with a single call or expression.

gorsing committed:
refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactor/irstate-attrs

refactoring

add pragma(inline, true)

change pragma(inline, true) -> pragma(inline, false)

introduce root.Array.only and test it out in expressionsem (#21520)

[attrib.d] remove unused semantic import

[dstruct.d] remove unused import

[aggregate.d] move `searchCtor` to `expressionsem.d`

remove some now unused semantic symbols from the import list.

[cond.d] remove unused `typesem` import (#21538)

[expression.d] restrict `typesem` import

fix #19587 No debug line info for simple code blocks (#21544)

set location info on return statement, not only on return expression

[dsymbol.d] remove dependance on glue layer (#21532)

move `loadModuleFromLibrary` to `dsymbolsem.d` (#21535)

remove `dmodule.d` dependance of a few more `dsymbolsem.d` symbols.

[func.d] remove dependance on glue layer (#21534)

[aggregate.d] move `checkOverlappedFields` to `dsymbolsem.d` (#21542)

and make private

Templatize `_adEq2` (#21513)

* Mark array literals as `on-stack` to bypass `@nogc` analysis

* Update `fail_compilation` test output

After updating the memcmp-ability criteria, now dynamic array can also
be compared using `memcmp` if their element type is comparable bit by
bit.

refactor: use new array constructor throughout expressionsem

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

ref

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactor/irstate-attrs

refactoring

add pragma(inline, true)

change pragma(inline, true) -> pragma(inline, false)

introduce root.Array.only and test it out in expressionsem (#21520)

[attrib.d] remove unused semantic import

[dstruct.d] remove unused import

[aggregate.d] move `searchCtor` to `expressionsem.d`

remove some now unused semantic symbols from the import list.

[cond.d] remove unused `typesem` import (#21538)

[expression.d] restrict `typesem` import

fix #19587 No debug line info for simple code blocks (#21544)

set location info on return statement, not only on return expression

[dsymbol.d] remove dependance on glue layer (#21532)

move `loadModuleFromLibrary` to `dsymbolsem.d` (#21535)

remove `dmodule.d` dependance of a few more `dsymbolsem.d` symbols.

[func.d] remove dependance on glue layer (#21534)

[aggregate.d] move `checkOverlappedFields` to `dsymbolsem.d` (#21542)

and make private

Templatize `_adEq2` (#21513)

* Mark array literals as `on-stack` to bypass `@nogc` analysis

* Update `fail_compilation` test output

After updating the memcmp-ability criteria, now dynamic array can also
be compared using `memcmp` if their element type is comparable bit by
bit.

refactor: use new array constructor throughout expressionsem

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

refactoring

ref

refactoring

refactoring

ref

refactoring

thewilsonator commented on compiler/src/dmd/toir.d:
```diff
@@ -106,15 +106,30 @@ struct IRState
      */
+    @property
+    @safe pure nothrow @nogc
+    pragma(inline, false)
```
do not mark this `pragma(inline, false)`

gorsing replied:
fixed

gorsing replied:
@thewilsonator thanks for your time and feedback

thewilsonator reviewed (commented):
I do not think this refactoring is desirable. `incUsageElem` and `arrayBoundsCheck` are neither large nor complicated function

thewilsonator reviewed (changes requested):
Again, what is the purpose of this PR? 
`arrayBoundsCheck` and `incUsageElem` are not large or complicated functions?
